src/util/sandbox/fd_sandbox_linux.c

```diff
@@ -0,0 +1,281 @@
+  profile->netns            = "fd-frank-netless",
+  profile->seccomp_prog     = NULL;
+  profile->user             = "nobody";
+  profile->close_fds_beyond = 3U;
+  profile->max_open_fds     = 3U;
+}
+
+int
+fd_sandbox( int * pargc,
+            char *** pargv,
+            fd_sandbox_profile_t profile ) {
+  
+  (void) pargc;
+  (void) pargv;
+
+  fd_sandbox_profile_linux_t linux_profile;
+  switch ( profile ) {
+    case FD_SANDBOX_PROFILE_COMMON:
+      fd_sandbox_profile_init_common( &linux_profile );
+    break;
+    default:
+    FD_LOG_ERR(( "unknown sandbox profile" ));
+  }
+  return fd_sandbox_linux( &linux_profile );
+}
+
+int
+fd_sandbox_linux( const fd_sandbox_profile_linux_t * profile ) {
+  if ( FD_UNLIKELY( !profile->initialized ) ) {
+    FD_LOG_ERR(( "the sandbox profile has not been initialized" ));
+  }
+
+  // Get target user before doing other kinds of restrictions
+  struct passwd * uinfo = getpwnam( profile->user );
+  if ( FD_UNLIKELY( !uinfo ) ) {
+    FD_LOG_ERR(( "getpwnam: %s", strerror( errno ) ));
+  }
+
+  // todo: maybe not clearenv but only keep vars prefixed with FD_
+  clearenv();
+  fd_sandbox_setup_netns( profile->netns );
+  fd_sandbox_setup_mountns( profile->chroot_path );
+  fd_sandbox_set_resource_limits( profile->max_open_fds );
+  fd_sandbox_change_user( uinfo );
+  fd_sandbox_close_fds_beyond( profile->close_fds_beyond );
+  fd_sandbox_drop_ambient_caps();
+  fd_sandbox_seccomp( profile->seccomp_prog );
+
+  FD_LOG_NOTICE(( "thread group sandboxed" ));
+  return 0;
+}
+
+void
+fd_sandbox_change_user( struct passwd * u ) {
+    
+  if ( FD_UNLIKELY( u->pw_uid == 0 ) ) {
+    FD_LOG_ERR(( "will not run as root" ));
+  }
+
+  FD_LOG_INFO(( "switching to user='%s' uid='%d' gid='%d'", u->pw_name, u->pw_uid, u->pw_gid ));
+
+  // setregid before setreuid otherwise setregid won't be allowed
+
+  if ( FD_UNLIKELY( setregid( u->pw_gid, u->pw_gid ) ) ) {
+    FD_LOG_ERR(( "setregid: %s", strerror( errno ) ));
+  }
+
+  if ( FD_UNLIKELY( setreuid( u->pw_uid, u->pw_uid ) ) ) {
+    FD_LOG_ERR(( "setreuid: %s", strerror( errno ) ));
+  }
+}
+
+void
+fd_sandbox_close_fds_beyond( uint target_max_fd ) {
+  FD_LOG_INFO(( "closing all fds beyond %d", target_max_fd ));
+  long max_fds = sysconf( _SC_OPEN_MAX );
+  for ( long fd = max_fds - 1; fd > target_max_fd; fd-- ) {
+     close( (int)fd );
+  } 
+}
+
+void
+fd_sandbox_set_resource_limits(uint max_open_fds) {
+  FD_LOG_INFO(( "setting resource limits" ));
+  struct rlimit l = {
+    .rlim_cur = max_open_fds,
+    .rlim_max = max_open_fds,
+  };
+
+  if ( FD_UNLIKELY( setrlimit(RLIMIT_NOFILE, &l) == -1 ) ) {
+    FD_LOG_ERR(( "setrlimit: %s", strerror( errno ) ));
+  }
+}
+
+void
+fd_sandbox_setup_netns( char * ns_name ) {
+  char netns_path[ PATH_MAX ];
+  int nsfd;
+
+  FD_LOG_INFO(( "setting up network namespace" ));
+
+  // realize the namespace path
+  int reslen = snprintf( netns_path, ARRAY_SIZE(netns_path), "/var/run/netns/%s", ns_name );
```

Comment:
Not sure why the network namespace needs to be named and set up / destoyed by the init script? I think you can just unshare a new NETNS here and have it be anonymous?

Reply:
You could decide to have it live in a preconfigured netns. It's probably what the RPC task is going to require.